File: lib/libtetrabz/example/dos.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "libtetrabz.h"
/**/
/*use libtetrabz, only : libtetrabz_fermieng, libtetrabz_dos*/
/**/
int main(int argc, char *argv[])
{
/**/
  int ltetra, nb, ng, nge[3], ngw[3], ii, i0, i1, i2, ik, ne, ie, nke, nkw;
  double bvec[9], ef, nelec, kvec[3], pi, sumdos;
  double *eig, *wght, *e0, *wght_dos;
  FILE *fp;
  /**/
  printf("Which tetrahedron method ?(1 = Linear, 2 = Optimized): ");
  scanf("%d", &ltetra);
  printf("\n k-point mesh ?: ");
  scanf("%d", &ng);
  /**/
  pi = acos(-1.0);
  nb = 1;
  for(ii = 0; ii < 3; ii++) nge[ii] = ng;
  for(ii = 0; ii < 3; ii++) ngw[ii] = ng;
  nelec = 0.5;
  bvec[0] = 1.0; bvec[1] = 0.0; bvec[2] = 0.0;
  bvec[3] = 0.0; bvec[4] = 1.0; bvec[5] = 0.0;
  bvec[6] = 0.0; bvec[7] = 0.0; bvec[8] = 1.0;
  nke = nge[0] * nge[1] * nge[2];
  nkw = ngw[0] * ngw[1] * ngw[2];
  /**/
  eig = (double*)malloc(nb * nke * sizeof(double));
  wght = (double*)malloc(nb * nkw * sizeof(double));
  /**/
  ik = 0;
  for(i2 = 0; i2 < nge[2]; i2++){
    for(i1 = 0; i1 < nge[1]; i1++){
      for(i0 = 0; i0 < nge[0]; i0++){
        /**/
        kvec[0] = 2.0 * pi * (double)(i0 - nge[0] / 2) / (double)nge[0];
        kvec[1] = 2.0 * pi * (double)(i1 - nge[1] / 2) / (double)nge[1];
        kvec[2] = 2.0 * pi * (double)(i2 - nge[2] / 2) / (double)nge[2];
        /**/
        eig[nb * ik] = - cos(kvec[0]) - cos(kvec[1]) - cos(kvec[2]);
        /**/
        ik += 1;
      }
    }
  }
  /**/
  libtetrabz_mp_libtetrabz_fermieng_(&ltetra,bvec,&nb,nge,eig,ngw,wght,&ef,&nelec);
  /**/
  printf("  E_F = %25.15e\n", ef);
  /**/
  ne = 100;
  wght_dos = (double*)malloc(ne * nb * nkw * sizeof(double));
  e0 = (double*)malloc(ne * sizeof(double));
  /**/
  for(ie = 0; ie < ne; ie++){
    e0[ie] = 6.0 / (double)(ne - 1) * (double)ie - 3.0;
  }
  /**/
  libtetrabz_mp_libtetrabz_dos_(&ltetra,bvec,&nb,nge,eig,ngw,wght_dos,&ne,e0);
  /**/
  fp = fopen("dos.dat", "w");
  for(ie = 0; ie < ne; ie++){
    /**/
    sumdos = 0.0;
    for(ik = 0; ik < nb * nkw; ik++) sumdos += wght_dos[ie + ik * ne];
    /**/
    fprintf(fp,"%25.15e  %25.15e\n", e0[ie], sumdos);
  }
  /**/
  fclose(fp);
  /**/
}

File: lib/libtetrabz/example/dos.f90
program dos
  !
  use libtetrabz, only : libtetrabz_fermieng, libtetrabz_dos
  implicit none
  !
  integer :: ltetra, nb, ng, nge(3), ngw(3), i1, i2, i3, ik, ne, ie, nke, nkw
  real(8) :: bvec(3,3), ef, nelec, kvec(3), pi
  real(8),allocatable :: eig(:,:), wght(:,:), e0(:), wght_dos(:,:,:)
  !
  write(*,'(a)', advance = "no") "Which tetrahedron method ?(1 = Linear, 2 = Optimized): "
  read(*,*) ltetra
  write(*,'(a)', advance = "no") "k-point mesh ?: "
  read(*,*) ng
  !
  pi = acos(-1d0)
  nb = 1
  nge(1:3) = ng
  ngw(1:3) = ng
  nelec = 0.5
  bvec(1:3,1) = (/1d0, 0d0, 0d0/)
  bvec(1:3,2) = (/0d0, 1d0, 0d0/)
  bvec(1:3,3) = (/0d0, 0d0, 1d0/)
  nke = product(nge(1:3))
  nkw = product(ngw(1:3))
  !
  allocate(eig(nb,nke), wght(nb,nkw))
  !
  ik = 0
  do i3 = 0, nge(3) - 1
     do i2 = 0, nge(2) - 1
        do i1 = 0, nge(1) - 1
           !
           ik = ik + 1
           kvec(1:3) = 2d0 * pi * dble((/i1, i2, i3/) - nge(1:3) / 2) / dble(nge(1:3))
           !
           eig(1,ik) = - sum(cos(kvec(1:3)))
           !
        end do
     end do
  end do
  !
  call libtetrabz_fermieng(ltetra,bvec,nb,nge,eig,ngw,wght,ef,nelec)
  !
  write(*,*) "  E_F = ", ef
  !
  ne = 100
  allocate(wght_dos(ne,nb,nkw), e0(ne))
  !
  do ie = 1, ne
     e0(ie) = 6d0 / dble(ne - 1) * dble(ie - 1) - 3d0
  end do
  !
  call libtetrabz_dos(ltetra,bvec,nb,nge,eig,ngw,wght_dos,ne,e0)
  !
  open(10, file = "dos.dat")
  do ie = 1, ne
     write(10,*) e0(ie), sum(wght_dos(ie,1:nb,1:nkw))
  end do
  !
  close(10)
  !
end program dos
